src/server/mod.rs

```diff
@@ -489,6 +512,28 @@ impl LanguageServer for LspServer {
                     .fold(value, |_acc, change| change.text.clone());
                 self.store.put(&key, &new_contents.clone());
                 self.publish_diagnostics(&key).await;
+
+                match self.state.lock() {
```

Comment:
We have added alot of locking. Most of the composition locks are fine, since those will be occurring less frequently. But the use of this lock in each keyDown (did_change)...makes me worried.

Reply:
`textDocument/didChange` doesn't fire with every keypress. It has some logic that might qualify as "debouncing." The lock isn't held any longer than it needs to be, and we can't _not_ have the lock, because rust wouldn't give us a bit of software then.

Also, we're already locking to update the server `Store` with each change. The API hides that fact (and I suspect that `store` and `state` will likely merge and do something similar.